public/client/windows/TracyETW.cpp

```diff
@@ -0,0 +1,248 @@
+
+struct Session {
+    EVENT_TRACE_PROPERTIES properties = {};
+    TCHAR name[64] = {};
+    CONTROLTRACE_ID handle = 0;
+};
+
+static ULONG ETWError(ULONG result) {
+    if (result == ERROR_SUCCESS)
+        return result;
+    ZoneScopedC(tracy::Color::Red4);
+    char message [128] = {};
+    size_t written = snprintf(message, sizeof(message), "ETW Error %u (0x%x): ", result, result);
+    written += FormatMessageA(
+        FORMAT_MESSAGE_FROM_SYSTEM | FORMAT_MESSAGE_IGNORE_INSERTS,
+        NULL,
+        result,
+        MAKELANGID(LANG_NEUTRAL, SUBLANG_DEFAULT),
+        (LPSTR)&message[written],
+        sizeof(message)-written,
+        NULL
+    );
+    tracy::InitCallstackCritical();
+    TracyMessageCS(message, written, tracy::Color::Red4, 60);
+#ifdef __cpp_exceptions
+    // TODO: should we throw an exception?
+#endif
+    return result;
+}
+
+static DWORD ElevatePrivilege(LPCTSTR PrivilegeName) {
+    TOKEN_PRIVILEGES tp = {};
+    tp.PrivilegeCount = 1;
+    tp.Privileges[0].Attributes = SE_PRIVILEGE_ENABLED;
+    if (LookupPrivilegeValue(nullptr, PrivilegeName, &tp.Privileges[0].Luid) == FALSE) {
+        return ETWError(GetLastError());
+    }
+    HANDLE hToken = {};
+    if (OpenProcessToken(GetCurrentProcess(), TOKEN_ADJUST_PRIVILEGES, &hToken) == FALSE) {
+        return ETWError(GetLastError());
+    }
+    BOOL adjusted = AdjustTokenPrivileges(hToken, FALSE, &tp, 0, nullptr, nullptr);
+    DWORD status = GetLastError();
+    CloseHandle(hToken);    // NOTE: skipping error handling for CloseHandle()
+    return ETWError(status);
+}
+
+static ULONG StartSession(Session& session) {
+    ULONG status = StartTrace(&session.handle, session.name, &session.properties);
+    if (status == ERROR_ALREADY_EXISTS) {
+        // Session is already running (likely from a previous run that did not terminate
+        // gracefully). There are two options: take control of the existing session with
+        // ControlSession(UPDATE), or stop the session and start fresh again. The latter
+        // is better because it also resets the event providers.
+        status = ControlTrace(session.handle, session.name, &session.properties, EVENT_TRACE_CONTROL_STOP);
+        if (status != ERROR_SUCCESS)
+            return ETWError(status);
+        status = StartTrace(&session.handle, session.name, &session.properties);
+    }
+    return ETWError(status);
+}
+
+static ULONG StopSession(Session& session) {
+    ULONG status = ControlTrace(session.handle, session.name, &session.properties, EVENT_TRACE_CONTROL_STOP);
+    if (status != ERROR_SUCCESS)
+        return ETWError(status);
+    // once stopped, the session handle becomes invalid
+    session.handle = 0;
+    return ERROR_SUCCESS;
+}
+
+static ULONG EnableProvider(
+    Session& session,
+    const GUID& ProviderId,
+    ULONG ControlCode = EVENT_CONTROL_CODE_ENABLE_PROVIDER,
+    UCHAR Level = TRACE_LEVEL_INFORMATION,
+    ULONGLONG MatchAnyKeyword = 0,
```

Comment:
Is `0` a good default, or should it be 0xFFFFFFFF ? Do we want also want `TRACE_LEVEL_INFORMATION`? I tend not to use default params to avoid such questions (personal preference)